lib/amazon_ssa_support/evm_queue.rb

```diff
@@ -0,0 +1,259 @@
+    #
+    def delete_request(req)
+      req[:sqs_msg].delete
+    end
+    
+    #################
+    # Reply methods
+    #################
+    
+    #
+    # Loop, reading extraction replies from extractors.
+    #
+    def get_reply_loop
+      @reply_queue.receive_messages.each do |msg|
+        next if (reply = get_reply(msg)).nil?
+        yield(reply)
+      end
+    end
+    
+    def get_reply(msg)
+      body = YAML.load(msg.body)
+      
+      case body[:reply_type]
+      when :extract
+        req_id = body[:request_id]
+        s3_obj_name = @reply_prefix + req_id
+        s3_obj = @reply_bucket.object(s3_obj_name)
+        unless s3_obj.exists?
+          $log.warn "#{self.class.name}.#{__method__}: Reply object #{s3_obj_name} does not exist"
+          msg.delete
+          return nil
+        end
+        reply_data = YAML.load(s3_obj.read)
+        reply_data[:request_id] = req_id
+        reply_data[:sqs_msg] = msg
+        s3_obj.delete
+        return reply_data
+      when :exit, :reboot, :shutdown
+        body[:sqs_msg] = msg
+        return body
+      else
+        $log.warn "#{self.class.name}.#{__method__}: Unrecognized reply type #{body[:reply_type]}"
+        return nil
+      end
+    end
+    private :get_reply
+    
+    #
+    # ers_reply = {
+    #  :reply_type    => :exit || :reboot || :shutdown
+    #  :extractor_id  => <The ID of the target of the request>
+    #  :request_id    => <The ID of the original request - not re-queued request>
+    # }
+    #
+    def send_ers_reply(req)
+      ers_reply  = {}
+      ers_reply[:reply_type]    = req[:request_type]
+      ers_reply[:extractor_id]  = @extractor_id
+      ers_reply[:request_id]    = req[:original_req_id] || req[:sqs_msg].message_id
+      
+      msg = @reply_queue.send_message(message_body: YAML.dump(ers_reply))
+      $log.info "#{self.class.name}.#{__method__}: sent reply (#{ers_reply[:reply_type]}) #{@reply_queue_name}:#{msg.message_id} to #{@request_queue_name}:#{ers_reply[:request_id]}"
+    end
+    
+    #
+    # Instantiate a new extract reply object for the extractor.
+    #
+    def new_reply(req)
+      SSAReply.new(req, self)
+    end
+    
+    #
+    # extract_reply = {
+    #  :reply_type    => :extract
+    #   :ec2_id      => <The ec2 id of the image/instance>,
+    #  :job_id      => <The id of the evm job requesting the extraction>,
+    #  :extractor_id  => <The id of the evm extractor instance performing the extract>
+    #   :start_time    => <The time the extraction started>,
+    #   :end_time    => <The time the extraction completed>,
+    #   :error      => <Error text and stack trace - if there was an error>,
+    #   :categories  => {
+    #     :accounts  => <XML text for accounts>,
+    #     :services  => <XML text for services>,
+    #     :software  => <XML text for software>,
+    #     :system    => <XML text for system>
+    #   }
+    # }
+    #
+    class SSAReply
+      def initialize(req, eeq)
```

Comment:
I don't like the "eeq" variable name.  This is an "EvmQueue" object - should be maybe "evmq" or "evm_q".